File: apps/minecraft/spigot/src/test/java/crabcraft/net/crabUtilities/PlayerVisibilityRegressionTest.java
package crabcraft.net.crabUtilities;

import org.bukkit.entity.Player;

import java.lang.reflect.Proxy;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public final class PlayerVisibilityRegressionTest {

    private PlayerVisibilityRegressionTest() {
    }

    public static void main(String[] args) {
        Player visible = player("VisiblePlayer", Set.of());
        Player hidden = player("HiddenPlayer", Set.of());
        Player viewer = player("Viewer", Set.of(hidden.getUniqueId()));

        List<Player> filtered = PlayerVisibility.visibleTo(
                viewer, List.of(visible, hidden));

        check(filtered.equals(List.of(visible)),
                "viewer-specific visibility did not remove the hidden player");
    }

    private static Player player(String name, Set<UUID> hidden) {
        UUID uuid = UUID.nameUUIDFromBytes(name.getBytes(StandardCharsets.UTF_8));
        return (Player) Proxy.newProxyInstance(
                Player.class.getClassLoader(),
                new Class<?>[]{Player.class},
                (proxy, method, args) -> switch (method.getName()) {
                    case "getName" -> name;
                    case "getUniqueId" -> uuid;
                    case "canSee" -> !hidden.contains(((Player) args[0]).getUniqueId());
                    case "equals" -> proxy == args[0];
                    case "hashCode" -> System.identityHashCode(proxy);
                    case "toString" -> name;
                    default -> defaultValue(method.getReturnType());
                });
    }

    private static Object defaultValue(Class<?> type) {
        if (!type.isPrimitive()) return null;
        if (type == boolean.class) return false;
        if (type == byte.class) return (byte) 0;
        if (type == short.class) return (short) 0;
        if (type == int.class) return 0;
        if (type == long.class) return 0L;
        if (type == float.class) return 0F;
        if (type == double.class) return 0D;
        if (type == char.class) return '\0';
        throw new IllegalArgumentException("Unsupported primitive: " + type);
    }

    private static void check(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }
}

File: apps/minecraft/spigot/src/test/kotlin/crabcraft/net/crabUtilities/PlayerVisibilityRegressionTest.kt
package crabcraft.net.crabUtilities

import java.lang.reflect.Proxy
import java.nio.charset.StandardCharsets
import java.util.UUID
import org.bukkit.entity.Player

object PlayerVisibilityRegressionTest {
    @JvmStatic
    fun main(args: Array<String>) {
        val visible = player("VisiblePlayer", emptySet())
        val hidden = player("HiddenPlayer", emptySet())
        val viewer = player("Viewer", setOf(hidden.uniqueId))
        check(
            PlayerVisibility.visibleTo(viewer, listOf(visible, hidden)) == listOf(visible),
            "viewer-specific visibility did not remove the hidden player",
        )
    }

    private fun player(name: String, hidden: Set<UUID>): Player {
        val uuid = UUID.nameUUIDFromBytes(name.toByteArray(StandardCharsets.UTF_8))
        return Proxy.newProxyInstance(Player::class.java.classLoader, arrayOf(Player::class.java)) { proxy, method, args
            ->
            when (method.name) {
                "getName" -> name
                "getUniqueId" -> uuid
                "canSee" -> (args!![0] as Player).uniqueId !in hidden
                "equals" -> proxy === args!![0]
                "hashCode" -> System.identityHashCode(proxy)
                "toString" -> name
                else -> defaultValue(method.returnType)
            }
        } as Player
    }

    private fun defaultValue(type: Class<*>): Any? =
        when (type) {
            java.lang.Boolean.TYPE -> false
            java.lang.Byte.TYPE -> 0.toByte()
            java.lang.Short.TYPE -> 0.toShort()
            java.lang.Integer.TYPE -> 0
            java.lang.Long.TYPE -> 0L
            java.lang.Float.TYPE -> 0F
            java.lang.Double.TYPE -> 0.0
            java.lang.Character.TYPE -> '\u0000'
            else -> if (!type.isPrimitive) null else throw IllegalArgumentException("Unsupported primitive: $type")
        }

    private fun check(condition: Boolean, message: String) {
        if (!condition) throw AssertionError(message)
    }
}
